### DISCUSSION TITLE
SU2 unexpected error while running shape_optimization.py

### DISCUSSION THREAD

**Discussion by LorenzoFabris** [su2code/SU2#1971] [Q&A]
**Describe the bug**
After defining the FFD box correctly, using SU2_DEF, the *shape_optimization.py* script was launched to perform adjoint optimization of an internal flow over airfoil.

The following error is printed out by SU2:
[SU2_error.txt](https://github.com/su2code/SU2/files/10999828/SU2_error.txt)

The .cfg file attached below is largely inspired by the Onera test case already published in the tutorial collection. 
However, I checked the code looking for some issues that my configuration might have caused and I couldn't find any (to my knowledge). It seems the solver has issues converting to float-type variable my text indications for DV_PARAM option. Moreover, I tried to "play" with the values related to DV_PARAM and no difference was noted.

The same issue was experienced both in version 7.5.0 and 7.5.1 regardless of running a pre-compiled version (Microsoft) or the built-from-source version (Ubuntu).

**To reproduce**

[airfoil.txt](https://github.com/su2code/SU2/files/10999881/airfoil.txt)
*Mesh file extension is not supported for upload*

**Bug report checklist**

- [ ✓] Consulting the build instructions (https://su2code.github.io/docs_v7/Build-SU2-Linux-MacOS/).
- [ ✓] Looking for similar problems on GitHub or CFD-Online (https://www.cfd-online.com/Forums/su2/).
- [ ✓] Updating to the newest version of SU2 (either develop, master, or the pre-built executables https://su2code.github.io/download.html).

**Desktop (please complete the following information):**
 - OS: Ubuntu 20.04.0 & Microsoft
 - C++ compiler and version: g++ 9.4.0
 - MPI implementation and version: Open MPI 4.0.3
 - SU2 Version: v7.5.1 & v7.5.0

*Thank you in advance for your help.*

**Comment by pcarruscag** [comment]
Try with FFD_CONTROL_POINT_2D

**Reply by LorenzoFabris** [reply]
> Try with FFD_CONTROL_POINT_2D

Hi Pedro, thank you for your quick reply. 
Initially, I didn't think about FFD_CONTROL_POINT_2D since my mesh is 3D. However, I applied your suggestion. Now the scripts starts the computation but another weird error comes out:

**Solver error**
[SU2_error2.txt](https://github.com/su2code/SU2/files/11004797/SU2_error2.txt)

Let me know if you have an explanation for it. 

For completeness, here attached the new .cfg file.
**Updated cfg file**
[airfoil.txt](https://github.com/su2code/SU2/files/11004802/airfoil.txt)

**Reply by pcarruscag** [reply]
Ok if your mesh is 3D you had the right setting. Maybe the python script has some issue dealing with only one design variable. Can you try with 2 points?

**Reply by LorenzoFabris** [reply]
> Ok if your mesh is 3D you had the right setting. Maybe the python script has some issue dealing with only one design variable. Can you try with 2 points?

I managed to start running the script with the following options, however it crashed after a while and I suspect because sesitivities calculated are zero. Instead, if I manually launch every script (SU2_CFD_AD, SU2_DOT, SU2_DEF. etc etc) I can complete the optimization procedure. However, I cannot iterate this process, obviously.

**CFG File**
[airfoil2.txt](https://github.com/su2code/SU2/files/11009765/airfoil2.txt)

**SU2 Error after crashing**
[SU2_error3.txt](https://github.com/su2code/SU2/files/11009792/SU2_error3.txt)

**Sensitivies calculated using shape_optimization.py**
![unnamed](https://user-images.githubusercontent.com/96534491/226143927-5a2fe54b-babd-4353-90c7-c02685abd77c.jpg)

**Sensitivities calculated manually operating the adjoint optimization**
![unnamed (1)](https://user-images.githubusercontent.com/96534491/226144176-c3cc558b-ceda-4559-9652-878895e621a9.jpg)


Last but not least, what do you mean by "trying with two points"? To which option for the design variables are you referring to? I'll be trying it immediately. Thank you!

**Comment by pcarruscag** [comment]
The ffd tag in DV param is wrong. I was just suggesting listing more points of the ffd as design parameters.
Honestly these optimization scripts are terrible, if it were up to me I would delete them all.
If you know how to manage the different SU2 components have a look at the tutorials that use FADO to build the optimization driver.

**Reply by LorenzoFabris** [reply]
> The ffd tag in DV param is wrong. I was just suggesting listing more points of the ffd as design parameters. Honestly these optimization scripts are terrible, if it were up to me I would delete them all. If you know how to manage the different SU2 components have a look at the tutorials that use FADO to build the optimization driver.

I corrected the FFD tag option. For sure my syntax was incorrect. However, if sensitivities are zero there must be something wrong considering my objective function is DRAG. 
Very curious aspect: if I run the constrained shape optimization of the OneraM6 (from tutorial collection) it works perfectly with the workflow I'm currently following. 

For completeness I'll attach here the current options and solver error:

**CFG File**
[airfoil2.txt](https://github.com/su2code/SU2/files/11009944/airfoil2.txt)

**SU2 Error**
[SU2_error4.txt](https://github.com/su2code/SU2/files/11009945/SU2_error4.txt)

What if I downgrade my SU2 version? Could it help? Maybe older versions present less issues (even if 3D-rotational periodicity was fixed by yourself in v7.3.1, hence for turbomachinery it might be an issue).
I know how to manage *some* of the SU2 components, I believe trying FADO could be helping in this issue. 

Please, let's keep in touch since my engineering department is fully focused on SU2 utilization and contribution and - even if not easy - communication between users and developers is important.
Thank you for any suggestion you'll provide to this comment and for your efforts.

**Reply by pcarruscag** [reply]
Note that when running SU2_CFD_AD you need to use SU2_DOT_AD, SU2_DOT should only be used with results of the continuous solver.
Have you tried applying some deformation via SU2_DEF? What are the gradients printed on screen when you run SU2_DOT_AD, different settings of FFD_CONTINUITY may help.

**Reply by LorenzoFabris** [reply]
> Note that when running SU2_CFD_AD you need to use SU2_DOT_AD, SU2_DOT should only be used with results of the continuous solver. Have you tried applying some deformation via SU2_DEF? What are the gradients printed on screen when you run SU2_DOT_AD, different settings of FFD_CONTINUITY may help.

I cannot apply any deformation with SU2_DEF since SU2_CFD_AD creashes after a while and so does the shape optimization script. I thought it was due to the value of the sensitivities, but I see here that those are non-zero (attached files).

### Here's the data I could produce with the latest *shape_optimization.py* run:

**Adjoint History**
[history_adjoint (copia).txt](https://github.com/su2code/SU2/files/11023658/history_adjoint.copia.txt)

**Surfaces Sensitivities**
[surface_sens (copia).txt](https://github.com/su2code/SU2/files/11023671/surface_sens.copia.txt)


### Here's the data I could produce with the latest *manual* run:

**Surface Sensitivities**
[surface_adjoint (copia).txt](https://github.com/su2code/SU2/files/11023682/surface_adjoint.copia.txt)

Other than the tutorial, isn't there a repository with some 3D-test cases to be fully reproduced?

Let me know if you have suggestions, thank you in advance.

**Reply by pcarruscag** [reply]
Yes but try running SU2_DEF directly with a nonzero value of deformation and check if you get sensible mesh deformation.
The adjoint solver is not converging properly. I would test the individual steps (DEF, CFD, CFD_AD, DOT_AD) manually and tweak the settings to make sure things are converging properly.

**Reply by LorenzoFabris** [reply]
> Yes but try running SU2_DEF directly with a nonzero value of deformation and check if you get sensible mesh deformation. The adjoint solver is not converging properly. I would test the individual steps (DEF, CFD, CFD_AD, DOT_AD) manually and tweak the settings to make sure things are converging properly.

Hi Pedro!
I ran the simulation all over again checking the single steps: 

0. **SU2_DEF** -> Generation of the FFD box
1. **SU2_CFD** -> Flow field solution
2. **SU2_CFD_AD** -> Adjoint flow simulation
3. **SU2_DOT_AD** -> Sensitivities calculation
4. **SU2_DEF** -> Mesh deformation (DV_VALUE=1.0)

For completeness:

**Surface flow sensitivities**
[surface_sens.txt](https://github.com/su2code/SU2/files/11057177/surface_sens.txt)

** Modified mesh**
[surface_deformed.txt](https://github.com/su2code/SU2/files/11057222/surface_deformed.txt)


Sensitivities are non-zero and the mesh generated is non other than a inifinitely "extruded" airfoil (DRAG is selected as obj. function, hence it is reasonable to expect this mesh modfication considering the mesh deformation is constrained). Also, the shape is slightly modified along flow direction.
I believe I might reach a better shape optimization changing the objective function or playing with the adjoint convergence parameters.

Moreover:

- Where does the drag come from? I'm running these simulations with Euler solver, therefore no viscous effects. Is numerical error calculated and then introduced as artifical viscosity (_discretization of convective terms e.g._)? 
- According to your experience, which are the parameters could be modified to efficiently reach an adjoint converged solution? My main issues derive after the flow field solution is reached.

Thank you for your help, let me know.

**Reply by pcarruscag** [reply]
Please do a search here or on CFD online for advice on how to improve the convergence.

**Reply by LorenzoFabris** [reply]
> Please do a search here or on CFD online for advice on how to improve the convergence.

Hi Pedro, thanks for you help a few days ago.
I'm running some simulations but what about this alert message?

------------------- Volumetric grid deformation (ZONE 0) ----------------
Performing the deformation of the volumetric grid.
CSysSolve::FGMRES(): system solved by initial guess.

WARNING: Convexity is not checked for 3D elements (issue #1171).


It is an old bug and in v7.5.0 I should not encounter it. Also my sideslip angle is 0.0.
Let me know if you know anything about this.
